ddtrace/tracer/stats.go

```diff
@@ -0,0 +1,332 @@
+		sb.data[s.key] = gs
+	}
+	if s.TopLevel {
+		gs.topLevelHits++
+	}
+	gs.hits++
+	if s.Error != 0 {
+		gs.errors++
+	}
+	gs.duration += uint64(s.Duration)
+	// alter resolution of duration distro
+	trundur := nsTimestampToFloat(s.Duration)
+	if s.Error != 0 {
+		gs.errDistribution.Add(trundur)
+	} else {
+		gs.okDistribution.Add(trundur)
+	}
+}
+
+// Export transforms a RawBucket into a statsBucket, typically used
+// before communicating data to the API, as RawBucket is the internal
+// type while statsBucket is the public, shared one.
+func (sb *rawBucket) Export() statsBucket {
+	csb := statsBucket{
+		Start:    sb.start,
+		Duration: sb.duration,
+		Stats:    make([]groupedStats, len(sb.data)),
+	}
+	for k, v := range sb.data {
+		b, err := v.export(k)
+		if err != nil {
+			log.Error("Could not export stats bucket: %v.", err)
+			continue
+		}
+		csb.Stats = append(csb.Stats, b)
+	}
+	return csb
+}
+
+type rawGroupedStats struct {
+	hits            uint64
+	topLevelHits    uint64
+	errors          uint64
+	duration        uint64
+	okDistribution  *ddsketch.DDSketch
+	errDistribution *ddsketch.DDSketch
+}
+
+func newRawGroupedStats() *rawGroupedStats {
+	const (
+		// relativeAccuracy is the value accuracy we have on the percentiles. For example, we can
+		// say that p99 is 100ms +- 1ms
+		relativeAccuracy = 0.01
+		// maxNumBins is the maximum number of bins of the ddSketch we use to store percentiles.
+		// It can affect relative accuracy, but in practice, 2048 bins is enough to have 1% relative accuracy from
+		// 80 micro second to 1 year: http://www.vldb.org/pvldb/vol12/p2195-masson.pdf
+		maxNumBins = 2048
+	)
+	okSketch, err := ddsketch.LogCollapsingLowestDenseDDSketch(relativeAccuracy, maxNumBins)
+	if err != nil {
+		log.Error("Error when creating ddsketch: %v", err)
+	}
+	errSketch, err := ddsketch.LogCollapsingLowestDenseDDSketch(relativeAccuracy, maxNumBins)
+	if err != nil {
+		log.Error("Error when creating ddsketch: %v", err)
+	}
+	return &rawGroupedStats{
+		okDistribution:  okSketch,
+		errDistribution: errSketch,
+	}
+}
+
+func (s *rawGroupedStats) export(k aggregation) (groupedStats, error) {
+	msg := s.okDistribution.ToProto()
+	okSummary, err := proto.Marshal(msg)
+	if err != nil {
+		return groupedStats{}, err
+	}
+	msg = s.errDistribution.ToProto()
+	errSummary, err := proto.Marshal(msg)
+	if err != nil {
+		return groupedStats{}, err
+	}
+	return groupedStats{
+		Service:        k.Service,
+		Name:           k.Name,
+		Resource:       k.Resource,
+		HTTPStatusCode: k.StatusCode,
+		Type:           k.Type,
+		Hits:           s.hits,
+		Errors:         s.errors,
+		Duration:       s.duration,
+		TopLevelHits:   s.topLevelHits,
+		OkSummary:      okSummary,
+		ErrorSummary:   errSummary,
+		Synthetics:     k.Synthetics,
+	}, nil
+}
+
+// nsTimestampToFloat converts a nanosec timestamp into a float nanosecond timestamp truncated to a fixed precision
+func nsTimestampToFloat(ns int64) float64 {
```

Comment:
why not divide by 1024 and multiply again?
Division by power of two should be really efficient.

Reply:
I copied this from the agent. I don't know what you mean. You'd have to spell (code) it out 🙂